doozerlib/runtime.py

```
@@ -644,56 +644,58 @@ def resolve_source(self, parent, meta):
-                    self.logger.info("Attempting to checkout source '%s' fallback-branch %s in: %s" % (alias, fallback_branch, source_dir))
-                    if fallback_branch != "master":
-                        rc2, out, err = exectools.cmd_gather(
-                            ["git", "checkout", "-b", fallback_branch, "origin/%s" % fallback_branch],
-                        )
-                    else:
-                        rc2 = 0
 
-                    if rc2 == 0:
-                        found = True
-                    else:
-                        self.logger.error("Failed checking out fallback-branch %s: %s" % (branch, err))
-                else:
-                    self.logger.error("Failed checking out branch %s: %s" % (branch, err))
+                self.logger.info('Target branch does not exist in {}, checking fallback branch {}'.format(url, fallback_branch))
+                rc, out, err = exectools.cmd_gather(branch_check.format(url, fallback_branch))
+                if rc:
+                    self.logger.error('Unable to check if fallback branch {} exists: {}'.format(fallback_branch, err))
+                    continue  # retry
+                if out.strip():
+                    clone_branch = fallback_branch
+                    break
 
-            if found:
-                # Store so that the next attempt to resolve the source hits the map
-                self.register_source_alias(alias, source_dir)
-                return source_dir
-            else:
-                raise DoozerFatalError("Error checking out target branch of source '%s' in: %s" % (alias, source_dir))
+        for _ in range(3):  # attempt checkout 3 times
+            try:
+                self.logger.info("Attempting to checkout source '%s' branch %s in: %s" % (url, clone_branch, source_dir))
+                exectools.cmd_assert("git clone -b {} --single-branch {} --depth 1 {}".format(clone_branch, url, source_dir))
```

Couldn't this reduce to:
```suggestion
                exectools.cmd_assert(
                  "git clone -b {} --single-branch {} --depth 1 {}".format(clone_branch, url, source_dir),
                  retries=3,
                  on_retry=["rm", "-rf", source_dir],
                )
```
Simpler to skip the for loop, try, and setting found; just let this fail as needed.

---

ack, 100% right... and I forgot the `rm -rf` between attempts. Good call.